celery/canvas.py

```diff
@@ -641,8 +652,13 @@ def stamp_links(self, visitor, **headers):
             visitor_headers = None
             if visitor is not None:
                 visitor_headers = visitor.on_callback(link, **headers) or {}
-            headers = self._stamp_headers(visitor_headers, **headers)
-            link.stamp(visitor=visitor, **headers)
+            headers = self._stamp_headers(
+                visitor_headers=visitor_headers,
+                append_stamps=append_stamps,
+                self_headers=False,  # Don't stamp links with own stamps implicitly
```

Comment:
The comment doesn't explain why that's the case

Reply:
I've added a better explanation - fix pushed.